README.md
# QMATIC Mobile Ticket
Qmatic mobile ticket solution is intended to issue e-tickets to the customers who are willing to go through the queue management solution by QMATIC. 
This is a web application that has been developed using Angular4 and the development environment of the project uses Angular-cli tools 
(https://github.com/angular/angular-cli).

## Prerequisites
* Angular/cli tools requires Node.js 8, or higher, together with NPM 3, or higher.
* API Gateway version 1.3.2.0 or higher required. Download from Qmatic portal.
* Supported Orchestra Versions
  * Orchestra 6.1 build 3.1.0.646 or later
  * Orchestra 6.2 build 3.2.0.289 or later

## Table of contents
- [Installation](#installation)
- [Setting up api gateway](#setting-up-api-gateway)
- [Proxy to Backend](#proxy-to-backend)
- [Running Unit Tests](#running-unit-tests)
- [Mobileticket.js library](#mobileticketjs-library)
- [Creating a Build](#creating-a-build)
- [Branding & Customization](#branding--customization)
- [Customer data](#customer-data)
- [Remote serving for customers](#remote-serving-for-customers)
- [Branch schedule](#branch-schedule)
- [One time password](#one-time-password)
- [Create ticket token](#create-ticket-token)
- [Cookie consent](#cookie-consent)
- [Set delay](#set-delay)
- [Set service group](#set-service-group)

## Installation
BEFORE YOU INSTALL: please read the [prerequisites](#prerequisites)

Clone the Mobile Ticket Solution
```
git clone https://github.com/qmatic/mobile-ticket.git
```
When the cloning is complete, install the required node modules by running the following commands from the project directory
```
npm install
cd node && npm install
```
We recommend Visual Studio Code (https://code.visualstudio.com/) as the IDE since it fits well with angular-cli tools. The original project is developed on visual code IDE.

NOTE: Earlier globally installed angular cli tools were necessary to run the project but it is no longer necessary.

## Setting up api gateway
Mobile ticket application uses Qmatic API gateway service as a proxy to talk to Orchestra. You have to download and install Qmatic API gate way service from [here](https://m01-qmaticworld.portal.qmatic.com/en/products/software/orchestra/api-gateway/software/#tabs).
Create a mobile user and generate a token by referring [API gateway manual](https://m01-qmaticworld.portal.qmatic.com/Documents/Qmatic%20World/Products/Software/Mobile%20Ticket/Manuals/224_03_F_MobileTicket.pdf). Copy the generated token to 'api_tokens' section in [service root folder]/conf/application.yml file.

It is required to change the API accesss token specified in the [project-root]/src/libs/js/mobileticket-{version}.js with the one you have generated as shown in the code snippet below. This is necessary only for the development, 
but in the production environment the auth token is read from the config file. So please refer to [Configuring the Proxy for Production Environment](#configuring-the-proxy-for-production-environment) to get more information 
on setting up auth-token for the production environment.

```js
  $.ajaxSetup({
    beforeSend: function (xhr) {
      xhr.setRequestHeader("Accept", "application/json");
      xhr.setRequestHeader("auth-token", "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"); // replace with your auth-token
    }
  });
```

## Orchestra configuration
For the Appointment Arrival feature it is needed that the Mobile user gets access to the Appointment and entrypoint connector.
To do this you need to create a role which contains the modules Appointment and Connector Entrypoint.

Then create a user with the name mobile and the same password as for the mobile user under System Administration > Paramaters.
Give this user access to all branches and the newly created Role.

## Proxy to Backend

Mobile Ticket solution is intended to work with QMATIC API Gateway service, which provides anonymous access to Orchestra REST API. The Mobile Ticket solution
is required to be hosted separately and all the REST API calls are proxied to avoid cross domain access violations. 
The current implementation is intended to be hosted on node.js and proxying is provided via express-http-proxy (https://www.npmjs.com/package/express-http-proxy).

#### Configuring ApiGateway for Development Environment

For development one needs to add the following five routes to the application.yml

```yml
    my_appointment:
        path: /MobileTicket/MyAppointment/find/*
        url: ${orchestra.central.url}/calendar-backend/api/v1/appointments/publicid
    central_appointment:        
        path: /MobileTicket/MyAppointment/findCentral/*
        url: ${orchestra.central.url}/qsystem/rest/appointment/appointments
    visit_app_entrypoint:
       path: /MobileTicket/MyAppointment/entrypoint/**
       url: ${orchestra.central.url}/qsystem/rest/entrypoint
    arrive_appointment:
       path: /MobileTicket/MyAppointment/arrive/**
       url: ${orchestra.central.url}/qsystem/rest/entrypoint
    meeting_info:
      path: /MobileTicket/MyMeeting/branches/**
      url: ${orchestra.central.url}/qsystem/rest/entrypoint/branches
    branch_schedule:
      path: /MobileTicket/BranchSchedule/**
      url: ${orchestra.central.url}/qsystem/rest/servicepoint
    sms_api:
      path: /rest/notification/**
      url: ${orchestra.central.url}/notification/
    delay_visit:
      path: /MobileTicket/DelayVisit/**
      url: ${orchestra.central.url}/qsystem/rest/entrypoint
    services_groups:
       path: /MobileTicket/servicesGroups/**  
       url: ${orchestra.central.url}/qsystem/rest/entrypoint/variables/servicesGroups
```
#### Configuring the Proxy for Development Environment

Edit proxy-config.json and set target to the IP and port of the QMATIC API Gateway service

File location
```html
project directory
|---proxy.config.json
```

```js
{
  "/MobileTicket/*": {
    "target": "http://192.168.1.35:9090",
    "secure": false
  },
  "/geo/*": {
    "target": "http://192.168.1.35:9090",
    "secure": false
  },
  "/MTService/*": {
    "target": "http://localhost:81",
    "secure": true
  }
}
```

If the API gateway is configures to use HTTPS the "target" parameters should indicate the protocol as "https".

If the certificate configured in API gateway has certificate errors or if it is a self signed certificate and still need the REST API calls to work with no issue, the setting "secure"
should be set to false. Having "secure" set to true will validate the certificate and REST calls will be failed if the certificate is not valid. The configurations for an API gateway configured
with a valid certificate will be as below.

```js
{
  "/MobileTicket/*": {
    "target": "https://192.168.1.35:9090",
    "secure": true
  },
  "/geo/*": {
    "target": "https://192.168.1.35:9090",
    "secure": true
  },
  "/MTService/*": {
    "target": "http://localhost:81",
    "secure": true
  }
}
```

Once the configuration is done you are ready to start the development server. You can open a console window in visual code and run npm start,
if you use visual code as the IDE. Otherwise just run this command on any console (bash, windows cli etc). Make sure you are in project root directory when running the command.
```
npm start
```
NOTE : npm start will run the command configured for "start" in package.json
```js
"scripts": {
    "start": "ng serve --proxy-config proxy.config.json",
    "lint": "tslint \"src/**/*.ts\"",
    "test": "ng test",
    "pree2e": "webdriver-manager update",
    "e2e": "protractor"
  }
```

#### Configuring the Proxy for Production Environment

Once the solution is built, the output folder structure contains the configuration file by the name proxy-config.json. Set the IP and port of QMATIC API Gateway service
to "value" in "apigw_ip_port". Then set a valid auth token to "value" in "auth_token".

If the API gateway is configured to use HTTPS the setting "gateway_has_certificate" must be set to true. 

If the certificate configured in API gateway has certificate errors or if it is a self signed certificate and still need the REST API calls to work with no issue, the setting "gateway_certificate_is_valid"
should be set to false. Having "gateway_certificate_is_valid" set to true will validate the certificate and REST calls will be failed if the certificate is not valid.

If the setting "embed_iFrame" set to true, application can embed in iFrame. For the securtiy purpose this setting set as false and can't embed in iFrame.

NOTE: WE DO NOT ENCOURAGE USING SELF SIGNED CERTIFICATES OR CERTIFICATES WITH ERRORS TO BE USED IN PRODUCTION ENVIRONMENT. THEREFORE WE RECOMMENT TO KEEP BOTH "gateway_has_certificate" and
"gateway_certificate_is_valid" SETTINGS TO SET TO true IN PRODUCTION ENVIRONMENT. 

File location
```html
project directory
|---dist
      |--- proxy-config.json
```

```js
{
    "apigw_ip_port": {
        "value": "10.2.3.88:9090",
        "description": "API Gateway - host address"
    },
    "local_webserver_port": {
        "value": "81",
        "description": "Local webserver port"
    },
    "auth_token": {
        "value": "nosecrets",
        "description": "Mobile API user - authentication token"
    },
    "local_webserver_ssl_port": {
        "value": "4443",
        "description": "Local webserver HTTPS port"
    },
    "support_ssl": {
        "value": "true",
        "description": "Flag indicating server supports SSL.Should have a valid public key and certificate in 'sslcert' folder "
    },
    "gateway_has_certificate": {
        "value": "true",
        "description": "Flag indicating the API gateway is configured to run on HTTPS"
    },
    "gateway_certificate_is_valid": {
        "value": "true",
        "description": "Flag indicating the configured SSL certificate in the API gateway is a valid. This setting is active only if 'gateway_has_certificate' is true. This must be set to 'false' if API gateway is configured with a self signed certificate "
    },
    "embed_iFrame": {
        "value": "false",
        "description": "Flag indicating the enable or disable X-FRAME-OPTIONS in http header to block iframe"
    },
    "tls_version": {
        "value": "",
        "description": "Flag indicating TLS version of the server. Possible values are 'TLSv1_method', 'TLSv1_1_method' and 'TLSv1_2_method'. Keep empty when using the latest version. Recommended for TLS 1.2 or later."
    },
    "hsts_expire_time": {
        "value": "63072000",
        "description": "Flag indicating HSTS(HTTP Strict-Transport-Security) expire time."
    },
    "cipher_set": {
        "value": [],
        "description": "Flag indicating cipher set which use in the server. Fill the value like this. ['ECDHE-RSA-AES128-SHA256', 'DHE-RSA-AES128-SHA256', 'AES128-GCM-SHA256']. Keep it empty when not going to modify."
    }
}
```
#### Changing the default configurations of the application in  Development/Production Environment
Application configuration file called 'config.json' contains the application specific parameters.

File location
```html
project directory
|---src
     |---app
           |---config
                  |---config.json
```

```js
{
    "branch_radius": {
        "value": "2147483647",
        "description": "Radius to include branches within (in meters)"
    },
    "appointment_arrive_radius": {
        "value": "2147483647",

release-notes/release-notes.md

<h1>Release notes Mobile Ticket 1</h1>

----------

<h2>Introduction</h2>

This document describes the new features, bug corrections, known issues and recommendations for Mobile Ticket 1. If you want to know about connector changes details or similar, this document is for you.

**Note:** Several of the remarks refer to a Jira number (Jira is Qmatic&#39;s internal registration system for bugs), or Pivotal Tracker (internal system for improvements and other issues).

<!--Add new update section after each release

<h2>Version UPDATE_VERSION_NUMBER</h2>

**Date:**
 
**Build number:**

<h3>Stories</h3>

| **Id** | **Release notes** |
| --- | --- |
| **xxx** | **Story header** Solution text |

<h3>Bug fixes</h3>

| **Id** | **Release notes** |
| --- | --- |
| **xxx** | **Bug header** Solution text |

<h3>Known issues</h3>

| **Id/Jira** | **Description** |
| --- | --- |
| **xxx** | **Bug header** Bug text |

<h3>Upgrade instructions</h3> 

----------
-->

<h2> Version 1.16.1 </h2>

**Date: 2022-03-07**
 
**Build number: 1**

<h3>Stories</h3>

| **Id** | **Release notes** |
| --- | --- |
| **MOB-306** | **Revisit the Issue fixed in MOB-271 after updating the angular version** |
| **MOB-780** | **Reimplement the handling of 503 for the function pollVisitStatus** |

----------

<h2> Version 1.16.0 </h2>

**Date: 2022-01-28**
 
**Build number: 5**

<h3>Stories</h3>

| **Id** | **Release notes** |
| --- | --- |
| **MOB-752** | **L&T Osnabruck - Mobile ticket with QR code with sammelkauf id and possibility to ask customer number.** |
| **MOB-760** | **Show more text in service group name list items** |
| **MOB-762** | **Include digicertglobalroot2 certificate in mobile ticket package** |
| **MOB-765** | **Update 'README' with 'npm -i instructions'** |

<h3>Upgrade Instructions</h3>

- When upgrading from a previous version, following should be considered.

1. 'additional_data' parameter under 'customer_data' needs to be added to config.json
```
 "customer_data": {
    "value": {
      "phone_number": {
        "value": "disable",
        "description": "Enable or disable customer phone number field in customer data section, 'enable => phone number field is visible but not mandatory', 'disable => phone number field is not visible' , 'mandatory => phone number field is visible and mandatory'"
      },
      "customerId": {
        "value": "disable",
        "description": "Enable or disable customer id field in customer data section, 'on = enable', 'off = disable'"
      },
      "additional_data": {
        "value": "",
        "description": "Enable or disable additonal customer data on QR scanning URL, set a value that is using in query parameter"
      }
    },
    "description": "Enable or disable customer id or phone number in customer area. If both are disabled customer data section will be hidden"
  }
```
----------

<h2> Version 1.15.0</h2>

**Date: 2021-11-10**
 
**Build number: 5**

<h3>Stories</h3>

| **Id** | **Release notes** |
| --- | --- |
| **MOB-634** | **"Powered by Qmatic" in Mobile Ticket** |
| **MOB-642** | **Remote Queueing capability in Mobile Ticket** |
| **MOB-730** | **Mobile Ticket - Service Grouping** |
| **MOB-720** | **Security improvements related to appointment id** |

<h3>Bug fixes</h3>

| **Id** | **Release notes** |
| --- | --- |
| **MOB-616** | **Prefilled mobile no in OTP screen get cleared if user open & close cookie consent popup** |
| **MOB-649** | **Text above ticket is not shown if the visit is called by a hardware device** |
| **MOB-648** | **Record all visits used Mobile Ticket application** |
| **MOB-668** | **windows server 2019 - Mobile Ticket cannot run as a service** |
| **MOB-661** | **Network error message doesn't appear if "show_queue_position" is disabled.** |
| **MOB-663** | **User can see queue position in the title tag, even if the "show_queue_position" is disabled.** |
| **MOB-704** | **Direct link to branch without locationservices shows different error.** |
| **MOB-705** | **Appointment Time - Text color should taken from theme-styles.css.** |
| **MOB-710** | **Cannot create mobileTicket from direct ticket url when both phn no & customer info are enabled** |
| **MOB-669** | **UI - Reduce the line height of button texts.** |

<h3>Upgrade Instructions</h3>

- When upgrading from a previous version, following should be considered.

1. 'footer' parameter needs to be added to config.json
```
 "footer": {
    "value": {
      "logo": {
        "value": "disable",
        "description": "show or hide the brand logo on the footer, enable = show, disable = hide"
      },
      "custom_text": {
        "value": "",
        "description": "show customize text on the footer, an empty value will hide the footer"
      }
    },
    "description": "Show or hide footer of the mobile ticket"
  }
```

2. 'delay_visit' parameter needs to be added to config.json
```
 "delay_visit": {
    "value": {
      "availability": {
        "value": "disable",
        "description": "Enable or disable whether customer need to create or update visit with delay"
      },
      "time_slot": {
        "value": [],
        "description": "Define delay time slots (minutes)"
      }
    },
    "description": "Enable or disable or define a time when customer need to create or update visit with delay"
  }
```

3. 'service_group' parameter needs to be added to config.json
```
 "service_group": {
    "value": {
      "availability": {
        "value": "enable",
        "description": "Enable or disable service grouping, 'on = enable', 'off = disable'"
      },
      "single_selection": {
        "value": "enable",

node/server.js
var proxy = require('express-http-proxy');
var express = require('express');
var compression = require('compression');
var zlib = require('zlib')
var expressStaticGzip = require("express-static-gzip");

var fs = require('fs');
var https = require('https');
const http = require('http');
var path = require('path');
var app = express();

var helmet = require('helmet');
var hidePoweredBy = require('hide-powered-by');
var csp = require('helmet-csp');
var nocache = require('nocache');
var nosniff = require('dont-sniff-mimetype');
var frameguard = require('frameguard');
var xssFilter = require('x-xss-protection');

var configFile = 'proxy-config.json';
var host = 'localhost:9090';
var authToken = 'nosecrets';
var port = '80';
var sslPort = '4443';
var supportSSL = false;
var validAPIGWCert = "1";
var APIGWHasSSL = true;
var isEmbedIFRAME = false;
var tlsVersion = '';
var hstsExpireTime = '63072000';
var tlsversionSet = ['TLSv1_method', 'TLSv1_1_method', 'TLSv1_2_method'];
var cipherSet = [];

var google_analytics = 'https://www.google-analytics.com';
var bootstarp_cdn = 'https://maxcdn.bootstrapcdn.com';

// Enable packet compression of each response
app.use(compression({level: zlib.Z_BEST_COMPRESSION, strategy: zlib.Z_DEFAULT_STRATEGY}));
// Set route to fetch compressed content
if (fs.existsSync('./src/zip')) {
   app.use("/zip", expressStaticGzip(__dirname + '/src/zip', {
	   enableBrotli: false,
    customCompressions: [{
        encodingName: "gzip",
        fileExtension: "gz"
    }]
   }));
}


//update configurations using config.json
var configuration = JSON.parse(
	fs.readFileSync(configFile)
);
// //update user-configurations using config.json for functional server
// var userConfiguration = JSON.parse(
// 	fs.readFileSync(userConfigFile)
// );

//Set well-known web vulnerabilities by setting HTTP headers appropriately
app.use(helmet());

//Hackers can exploit known vulnerabilities in Express/Node if they see that your site is 
//powered by Express (or whichever framework you use)
app.use(hidePoweredBy());

//Abolish all JS client-side caching.
app.use(nocache());

//Browsers will try to "sniff" mimetypes. For example, if my server serves file.txt with a text/plain content-type, 
//some browsers can still run that file with <script src="file.txt"></script>
app.use(nosniff());

//The X-Frame-Options HTTP header restricts who can put your site in a frame
// action either can be {action:'deny'}, {action:'sameorigin'}, {action: 'allow-from', domain: 'http://example.com'}
app.use(frameguard({ action: 'deny' }));

//The X-XSS-Protection HTTP header is a basic protection against XSS
app.use(xssFilter());

//Content Security Policy helps prevent unwanted content being injected into your webpages
app.use(csp({
	directives: {
	defaultSrc: ["'self'"],
	scriptSrc: ["'self'", "'unsafe-inline'", "'unsafe-eval'",google_analytics ],
	styleSrc: ["'self'", "'unsafe-inline'",bootstarp_cdn],
	fontSrc: ["'self'", 'data:',bootstarp_cdn],
	imgSrc: ["'self'", 'data:', google_analytics],
	sandbox: ['allow-forms', 'allow-scripts', 'allow-same-origin', 'allow-popups'],
	reportUri: '/report-violation',
	objectSrc: ["'none'"]
	},

	// This module will detect common mistakes in your directives and throw errors
	// if it finds any. To disable this, enable "loose mode".
	loose: false,

	// Set to true if you only want browsers to report errors, not block them.
	// You may also set this to a function(req, res) in order to decide dynamically
	// whether to use reportOnly mode, e.g., to allow for a dynamic kill switch.
	reportOnly: false,

	// Set to true if you want to blindly set all headers: Content-Security-Policy,
	// X-WebKit-CSP, and X-Content-Security-Policy.
	setAllHeaders: false,

	// Set to true if you want to disable CSP on Android where it can be buggy.
	disableAndroid: false,

	// Set to false if you want to completely disable any user-agent sniffing.
	// This may make the headers less compatible but it will be much faster.
	// This defaults to `true`.
	browserSniff: true
}));

host = configuration.apigw_ip_port.value;
port = configuration.local_webserver_port.value;
authToken = configuration.auth_token.value;
sslPort = configuration.local_webserver_ssl_port.value;
supportSSL = (configuration.support_ssl.value.trim() === 'true')?true:false;
validAPIGWCert = (configuration.gateway_certificate_is_valid.value.trim()=== 'true')?"1":"0";
APIGWHasSSL = (configuration.gateway_has_certificate.value.trim()=== 'true')?true:false;
isEmbedIFRAME = (configuration.embed_iFrame.value.trim() === 'true')?true:false;
tlsVersion = configuration.tls_version.value;
hstsExpireTime = configuration.hsts_expire_time.value;
cipherSet = configuration.cipher_set.value;

// // functional server services
// otpService = userConfiguration.otp_service.value;
// fsPort = configuration.local_functional_server_port.value;

//this will bypass certificate errors in node to API gateway encrypted channel, if set to '1'
//if '0' communication will be blocked. So production this should be set to '0'
process.env.NODE_TLS_REJECT_UNAUTHORIZED = validAPIGWCert;

var privateKey, certificate, credentials;

if (supportSSL) {
	privateKey = fs.readFileSync('sslcert/server.key', 'utf8');
	certificate = fs.readFileSync('sslcert/server.crt', 'utf8');

	credentials = { key: privateKey, cert: certificate };
	if (tlsVersion.length > 0 && tlsversionSet.indexOf(tlsVersion) !== -1) {
		credentials.secureProtocol = tlsVersion;
	} 
	if (cipherSet.length > 0) {
		credentials.ciphers = cipherSet.join(':');
	}
}

var options = {
	setHeaders: function (res, path, stat) {
		res = handleHeaders(res);
	}
}

app.use(express.static(__dirname + '/src', options));


// Redirect no_branch requests to index.html
app.get('/no_support$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect no_branch requests to index.html
app.get('/no_branch$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect fingerprint
app.get('/fingerprintjs2$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'fingerprint2.js'));
});
// Redirect no_visit requests to index.html
app.get('/no_visit$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect all requests that start with branches and end, to index.html
app.get('/branches*', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect all requests that start with services and end, to index.html
app.get('/services$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect all requests that start with ticket info and end, to index.html
app.get('/ticket$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect all requests that start with branches and end, to index.html
app.get('/open_hours$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});
// Redirect all requests that start with appointment and end, to index.html
app.get('/appointment$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect all requests that start with customer data and end, to index.html
app.get('/customer_data$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

// Redirect all requests that start with privacy policy and end, to index.html
app.get('/privacy_policy$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

app.get('/cookie_consent$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

app.get('/otp_number$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});

app.get('/otp_pin$', function (req, res) {
	res = handleHeaders(res);
  	res.sendFile(path.join(__dirname + '/src', 'index.html'));
});


// Proxy mobile example to API gateway
var apiProxy = proxy(host, {
	// ip and port off apigateway

	proxyReqPathResolver: function (req) {
		return require('url').parse(req.originalUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL
});

var apiFindProxy = proxy(host, {	// ip and port off apigateway
	proxyReqPathResolver: function (req) {
		var newUrl = req.originalUrl.replace("/MobileTicket/MyAppointment/find/","/rest/calendar-backend/api/v1/appointments/publicid/");
		return require('url').parse(newUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL,
	userResDecorator: function(proxyRes, proxyResData, userReq, userRes) {
		data = JSON.parse(proxyResData.toString('utf8'));
		newData = {};
		newData.appointment = {};
		if (data.appointment !== undefined) {
			newData.appointment.qpId = data.appointment.qpId
			newData.appointment.branch = data.appointment.branch;
			newData.appointment.start = data.appointment.start;
			newData.appointment.services = data.appointment.services;
		}
		return JSON.stringify(newData);
	}
});

var apiFindCentralProxy = proxy(host, {	// ip and port off apigateway
	proxyReqPathResolver: function (req) {
		var newUrl = req.originalUrl.replace("/MobileTicket/MyAppointment/findCentral/","/rest/appointment/appointments/");
		return require('url').parse(newUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL,
	userResDecorator: function(proxyRes, proxyResData, userReq, userRes) {
		data = JSON.parse(proxyResData.toString('utf8'));
		newData = {};
		if (data !== undefined) {
			newData.services = data.services;
			newData.status = data.status
			newData.branchId = data.branchId;
			newData.startTime = data.startTime;
			newData.endTime = data.endTime;
			newData.properties = {};
			newData.properties.notes = data.properties.notes;
		}
		return JSON.stringify(newData);
	}
});

var apiEntryPointProxy = proxy(host, {
	// ip and port off apigateway

	proxyReqPathResolver: function (req) {
		var newUrl = req.originalUrl.replace("/MobileTicket/MyAppointment/entrypoint/branches/","/rest/entrypoint/branches/");
		return require('url').parse(newUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL
});

var apiArriveProxy = proxy(host, {
	// ip and port off apigateway

	proxyReqPathResolver: function (req) {
		var newUrl = req.originalUrl.replace("/MobileTicket/MyAppointment/arrive/branches/","/rest/entrypoint/branches/");
		return require('url').parse(newUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL,
	userResDecorator: function(proxyRes, proxyResData, userReq, userRes) {
		data = JSON.parse(proxyResData.toString('utf8'));
		newData = {};
		newData.parameterMap = {};
		if (data !== undefined) {
			newData.id = data.id;
			newData.ticketId = data.ticketId;
			newData.currentVisitService = data.currentVisitService;
			newData.checksum = data.checksum;
			newData.parameterMap.startQueueOrigId = data.parameterMap.startQueueOrigId;
			newData.parameterMap.branchName = data.parameterMap.branchName;
		}
		return JSON.stringify(newData);
	}
});

var apiMeetingProxy = proxy(host, {
	// ip and port off apigateway

	proxyReqPathResolver: function (req) {
		var newUrl = req.originalUrl.replace("/MobileTicket/MyMeeting/branches/","/rest/entrypoint/branches/");
		return require('url').parse(newUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL,
	userResDecorator: function(proxyRes, proxyResData, userReq, userRes) {
		data = JSON.parse(proxyResData.toString('utf8'));
		newData = {};
		newData.parameterMap = {};
		if (data !== undefined) {
			newData.ticketId = data.ticketId;
			newData.checksum = data.checksum;
			newData.parameterMap.meetingUrl = data.parameterMap.meetingUrl;
			
		}
		return JSON.stringify(newData);
	}
});

// proxy for MT service
var apiOtpProxy = proxy('localhost:82', {
	// ip and port off apigateway
	proxyReqPathResolver: function (req) {
		console.log(req.originalUrl);
		return require('url').parse(req.originalUrl).path;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		console.log(123);
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL
});
var apiButtonSheduleProxy = proxy(host, {
	// ip and port off apigateway

	proxyReqPathResolver: function (req) {
		var newUrl = req.originalUrl.replace("/MobileTicket/BranchShedule/","/rest/servicepoint/");
		return require('url').parse(newUrl).path;
	},
	proxyReqOptDecorator: function(proxyReqOpts, srcReq) {
		proxyReqOpts.headers['auth-token'] = authToken		// api_token for mobile user
		proxyReqOpts.headers['Content-Type'] = 'application/json'
		return proxyReqOpts;
	},
	userResHeaderDecorator(headers, userReq, userRes, proxyReq, proxyRes) {
		if (isEmbedIFRAME === false) {
			headers['X-Frame-Options'] = "DENY";
		}
		headers['Content-Security-Policy'] = "default-src \'self\'";
	
		if (supportSSL) {
			headers['Strict-Transport-Security'] = "max-age=" + hstsExpireTime + "; includeSubDomains";
		}
		return headers;
	},
	https: APIGWHasSSL
});

var handleHeaders = function (res) {
	if (isEmbedIFRAME === false) {
		res.set('X-Frame-Options', "DENY");
	} else {
		res.removeHeader('X-Frame-Options');
	}

	if (supportSSL) {
		res.set('Strict-Transport-Security', "max-age=" + hstsExpireTime +"; includeSubDomains");
	}
	return res;
}

app.post('/notification/*', (req, res) => {
	console.log('post hit');
});

app.use("/geo/branches/*", apiProxy);
app.use("/MobileTicket/branches/*", apiProxy);
app.use("/MobileTicket/MyAppointment/find/*", apiFindProxy);
app.use("/MobileTicket/MyAppointment/findCentral/*", apiFindCentralProxy);
app.use("/MobileTicket/MyAppointment/entrypoint/*", apiEntryPointProxy);
app.use("/MobileTicket/MyAppointment/arrive/*", apiArriveProxy);
app.use("/MobileTicket/services/*", apiProxy);
app.use("/MobileTicket/MyVisit/*", apiProxy);
app.use("/MobileTicket/MyMeeting/*", apiMeetingProxy);
app.use("/MTService/*", apiOtpProxy);
app.use("/MobileTicket/BranchShedule/*", apiButtonSheduleProxy);


if (supportSSL) {
	var httpsServer = https.createServer(credentials, app);
	httpsServer.listen(sslPort, function () {
		var listenAddress = httpsServer.address().address;
		var listenPort = httpsServer.address().port;

		console.log("Mobile Ticket app listening at https://%s:%s over SSL", listenAddress, listenPort);
	});
} else{
	var server = app.listen(port, function () {  // port the mobileTicket will listen to.
	var listenAddress = server.address().address;
	var listenPort = server.address().port;

	console.log("Mobile Ticket app listening at http://%s:%s", listenAddress, listenPort);
});
}

